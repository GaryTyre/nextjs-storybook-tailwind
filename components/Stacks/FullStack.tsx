import { useState } from 'react'
import { MenuAlt2Icon } from '@heroicons/react/outline'
import TopMenu from './TopMenu/TopMenu'
import Breadcrumbs from './Breadcrumbs'
import SideMenu from './SideMenu/SideMenu'
import SideMenuMobile from './SideMenu/SideMenuMobile'
import { ContentArea } from './ContentArea'

export type FullStackProps = {
  children: React.ReactNode;
}

export const FullStack: React.FC<FullStackProps> = ({
  children
}) => {
  const [mobileMenuOpen, setMobileMenuOpen] = useState(false);

  const closeMobileSideMenu = () => {
    setMobileMenuOpen(false);
  };

  return (
    <>
      <TopMenu></TopMenu>
      <div className="h-full flex">
        <SideMenu></SideMenu>
        <SideMenuMobile isOpen={mobileMenuOpen} closeMenu={closeMobileSideMenu}></SideMenuMobile>

        <div className="flex-1 flex flex-col overflow-hidden">
          <header className="w-full">
            <div className="relative z-10 flex-shrink-0 h-12 bg-white border-b border-gray-200 shadow-sm flex">
              <button
                type="button"
                className="border-r border-gray-200 px-4 text-gray-500 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-gray-500 md:hidden"
                onClick={() => setMobileMenuOpen(true)}
              >
                <span className="sr-only">Open sidebar</span>
                <MenuAlt2Icon className="h-6 w-6" aria-hidden="true" />
              </button>
              <Breadcrumbs></Breadcrumbs>
            </div>
          </header>

          <ContentArea>
            { children }
          </ContentArea>
        </div>
      </div>
    </>
  )
}

export default FullStack;