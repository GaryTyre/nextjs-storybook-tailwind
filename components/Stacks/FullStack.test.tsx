/* eslint-env jest */
import React from "react";
import { render, screen, waitFor } from '@testing-library/react';
import { findByText } from '@testing-library/dom'
import "@testing-library/jest-dom/extend-expect";
//ðŸ‘‡ Imports a specific story for the test
import { Default } from "./FullStack.stories";
import userEvent from '@testing-library/user-event';
// import { sampleMenu } from "../../utils/test-menu-utils";
// import { SideMenuContext } from "../../contexts/SideMenuContext";

describe("FullStack", () => {
  it("renders the component", () => {
    const tree = render(<Default {...Default.args} />);
    expect(tree.container).toMatchSnapshot();
  });
  it("renders the component with the correct children ", async () => {
   const container = render(<Default {...Default.args} ><p id="stack">STACK</p></Default>);
   expect(await findByText(container.container, 'STACK')).toBeVisible();
  });
  // it("renders the component with the correct menu items ", async () => {
  //  const container = render(<Default {...Default.args} menuItems={sampleMenu} />);
  // sampleMenu.forEach(async menuItem => {
  //   expect(await findByText(container.container, `${menuItem}`)).toBeVisible();
  // })
  // });
  test("Should call the supplied logout handler", async () => {
    const logoutHandler = jest.fn();
    const tree = render(<Default {...Default.args} logout={logoutHandler} />);
    const btn = screen.getByRole("button", {
      name: /Logout/i,
    });
    userEvent.click(btn);
    await waitFor(() => expect(logoutHandler).toHaveBeenCalled());
    expect(tree.container).toMatchSnapshot();
  });
  it('user can see and click a button to open the sidemenu (calls SideMenuContext.sideMenu.toggleOpen', async () => {
    const sideMenu =  { 
      open: false, 
      toggleOpen: jest.fn(),
     }; 
  //   <SideMenuContext.Provider value={{ sideMenu: sideMenu}}>
  //   <Default {...Default.args} />
  // </SideMenuContext.Provider>
    const buttonHandler = jest.fn();
    const tree = render(<Default {...Default.args} />)
    const btn = screen.getByRole("button", {
      name: /Open sidebar/i,
    });
    userEvent.click(btn);
    await waitFor(() => expect(sideMenu.toggleOpen).toHaveBeenCalled())
    expect(tree.container).toMatchSnapshot()
  });
});
