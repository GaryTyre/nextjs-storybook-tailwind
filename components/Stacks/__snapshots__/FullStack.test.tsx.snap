// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullStack renders the component 1`] = `
<div>
  <div
    class="relative bg-white"
  >
    <div
      class="relative z-20"
    >
      <div
        aria-hidden="true"
        class="absolute inset-0 shadow pointer-events-none"
      />
      <div
        class="flex items-center justify-between w-full px-4 py-5 mx-auto sm:px-6 sm:py-4 lg:px-8 md:justify-start md:space-x-10"
      >
        <div>
          <a
            class="flex"
            href="#"
          >
            <span
              class="sr-only"
            >
              Workflow
            </span>
            <img
              alt=""
              class="w-auto h-8 sm:h-10"
              src="https://media-exp1.licdn.com/dms/image/C510BAQHThKCx51J1GQ/company-logo_200_200/0/1519870029475?e=1651104000&v=beta&t=xfUhWsU-_P0O1P0hHo3YV_JXIvZX8wqgmTstBJMp9KM"
            />
          </a>
        </div>
        <div
          class="-my-2 -mr-2 md:hidden"
        >
          <button
            aria-expanded="false"
            class="inline-flex items-center justify-center p-2 text-gray-400 bg-white rounded-md hover:text-gray-500 hover:bg-gray-100 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-indigo-500"
            id="headlessui-popover-button-4"
            type="button"
          >
            <span
              class="sr-only"
            >
              Open menu
            </span>
            <svg
              aria-hidden="true"
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
        <div
          class="hidden md:flex-1 md:flex md:items-center md:justify-between"
        >
          <nav
            class="flex space-x-10"
          >
            <div>
              <button
                aria-expanded="false"
                class="text-gray-500 group bg-white rounded-md inline-flex items-center text-base font-medium py-2 px-1 hover:text-gray-900 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-800"
                id="headlessui-popover-button-1"
                type="button"
              >
                <span>
                  Supplier Manager
                </span>
                <svg
                  aria-hidden="true"
                  class="text-gray-400 ml-2 h-5 w-5 group-hover:text-gray-500"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </nav>
          <div
            class="relative ml-3"
          >
            <div>
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="flex items-center max-w-xs text-sm bg-white rounded-full focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
                id="headlessui-menu-button-3"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  Open user menu
                </span>
                <img
                  alt=""
                  class="w-8 h-8 rounded-full"
                  src="https://images.unsplash.com/photo-1472099645785-5658abf4ff4e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=facearea&facepad=2&w=256&h=256&q=80"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="flex h-full"
  >
    <div
      class="hidden w-28 bg-gray-700 overflow-y-auto md:block"
    >
      <div
        class="w-full py-2 flex flex-col items-center"
      >
        <div
          class="flex-1 w-full px-2 space-y-1"
        >
          <a
            aria-current="page"
            class="bg-gray-800 text-white group w-full p-3 rounded-md flex flex-col items-center text-xs font-medium text-center"
            href="#"
          >
            <svg
              aria-hidden="true"
              class="text-white h-8 w-8"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 11H5m14 0a2 2 0 012 2v6a2 2 0 01-2 2H5a2 2 0 01-2-2v-6a2 2 0 012-2m14 0V9a2 2 0 00-2-2M5 11V9a2 2 0 012-2m0 0V5a2 2 0 012-2h6a2 2 0 012 2v2M7 7h10"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="mt-2"
            >
              Supplier List
            </span>
          </a>
          <a
            class="text-gray-100 hover:bg-gray-800 hover:text-white group w-full p-3 rounded-md flex flex-col items-center text-xs font-medium text-center"
            href="#"
          >
            <svg
              aria-hidden="true"
              class="text-gray-300 group-hover:text-white h-8 w-8"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6a2 2 0 012-2h2a2 2 0 012 2v2a2 2 0 01-2 2H6a2 2 0 01-2-2V6zM14 6a2 2 0 012-2h2a2 2 0 012 2v2a2 2 0 01-2 2h-2a2 2 0 01-2-2V6zM4 16a2 2 0 012-2h2a2 2 0 012 2v2a2 2 0 01-2 2H6a2 2 0 01-2-2v-2zM14 16a2 2 0 012-2h2a2 2 0 012 2v2a2 2 0 01-2 2h-2a2 2 0 01-2-2v-2z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="mt-2"
            >
              Customer
Approved List
            </span>
          </a>
          <a
            class="text-gray-100 hover:bg-gray-800 hover:text-white group w-full p-3 rounded-md flex flex-col items-center text-xs font-medium text-center"
            href="#"
          >
            <svg
              aria-hidden="true"
              class="text-gray-300 group-hover:text-white h-8 w-8"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 13l4 4L19 7"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="mt-2"
            >
              Supplier Approval
            </span>
          </a>
          <a
            class="text-gray-100 hover:bg-gray-800 hover:text-white group w-full p-3 rounded-md flex flex-col items-center text-xs font-medium text-center"
            href="#"
          >
            <svg
              aria-hidden="true"
              class="text-gray-300 group-hover:text-white h-8 w-8"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="mt-2"
            >
              Compliance
            </span>
          </a>
          <a
            class="text-gray-100 hover:bg-gray-800 hover:text-white group w-full p-3 rounded-md flex flex-col items-center text-xs font-medium text-center"
            href="#"
          >
            <svg
              aria-hidden="true"
              class="text-gray-300 group-hover:text-white h-8 w-8"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 20h5v-2a3 3 0 00-5.356-1.857M17 20H7m10 0v-2c0-.656-.126-1.283-.356-1.857M7 20H2v-2a3 3 0 015.356-1.857M7 20v-2c0-.656.126-1.283.356-1.857m0 0a5.002 5.002 0 019.288 0M15 7a3 3 0 11-6 0 3 3 0 016 0zm6 3a2 2 0 11-4 0 2 2 0 014 0zM7 10a2 2 0 11-4 0 2 2 0 014 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="mt-2"
            >
              Contacts
            </span>
          </a>
          <a
            class="text-gray-100 hover:bg-gray-800 hover:text-white group w-full p-3 rounded-md flex flex-col items-center text-xs font-medium text-center"
            href="#"
          >
            <svg
              aria-hidden="true"
              class="text-gray-300 group-hover:text-white h-8 w-8"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 20l-5.447-2.724A1 1 0 013 16.382V5.618a1 1 0 011.447-.894L9 7m0 13l6-3m-6 3V7m6 10l4.553 2.276A1 1 0 0021 18.382V7.618a1 1 0 00-.553-.894L15 4m0 13V4m0 0L9 7"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="mt-2"
            >
              Maps
            </span>
          </a>
          <a
            class="text-gray-100 hover:bg-gray-800 hover:text-white group w-full p-3 rounded-md flex flex-col items-center text-xs font-medium text-center"
            href="#"
          >
            <svg
              aria-hidden="true"
              class="text-gray-300 group-hover:text-white h-8 w-8"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 5c7.18 0 13 5.82 13 13M6 11a7 7 0 017 7m-6 0a1 1 0 11-2 0 1 1 0 012 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="mt-2"
            >
              Activity
            </span>
          </a>
          <a
            class="text-gray-100 hover:bg-gray-800 hover:text-white group w-full p-3 rounded-md flex flex-col items-center text-xs font-medium text-center"
            href="#"
          >
            <svg
              aria-hidden="true"
              class="text-gray-300 group-hover:text-white h-8 w-8"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10.325 4.317c.426-1.756 2.924-1.756 3.35 0a1.724 1.724 0 002.573 1.066c1.543-.94 3.31.826 2.37 2.37a1.724 1.724 0 001.065 2.572c1.756.426 1.756 2.924 0 3.35a1.724 1.724 0 00-1.066 2.573c.94 1.543-.826 3.31-2.37 2.37a1.724 1.724 0 00-2.572 1.065c-.426 1.756-2.924 1.756-3.35 0a1.724 1.724 0 00-2.573-1.066c-1.543.94-3.31-.826-2.37-2.37a1.724 1.724 0 00-1.065-2.572c-1.756-.426-1.756-2.924 0-3.35a1.724 1.724 0 001.066-2.573c-.94-1.543.826-3.31 2.37-2.37.996.608 2.296.07 2.572-1.065z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
              <path
                d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="mt-2"
            >
              Admin
            </span>
          </a>
        </div>
      </div>
    </div>
    <div
      class="flex flex-col flex-1 overflow-hidden"
    >
      <header
        class="w-full"
      >
        <div
          class="relative z-10 flex flex-shrink-0 h-12 bg-white border-b border-gray-200 shadow-sm"
        >
          <button
            class="px-4 text-gray-500 border-r border-gray-200 focus:outline-none focus:ring-2 focus:ring-inset focus:ring-gray-500 md:hidden"
            type="button"
          >
            <span
              class="sr-only"
            >
              Open sidebar
            </span>
            <svg
              aria-hidden="true"
              class="w-6 h-6"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h7"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
          <nav
            aria-label="Breadcrumb"
            class="flex ml-5"
          >
            <ol
              class="flex items-center space-x-4"
              role="list"
            >
              <li>
                <div>
                  <a
                    class="text-gray-300 hover:text-gray-400"
                    href="#"
                  >
                    <svg
                      aria-hidden="true"
                      class="flex-shrink-0 w-5 h-5"
                      fill="currentColor"
                      viewBox="0 0 20 20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M10.707 2.293a1 1 0 00-1.414 0l-7 7a1 1 0 001.414 1.414L4 10.414V17a1 1 0 001 1h2a1 1 0 001-1v-2a1 1 0 011-1h2a1 1 0 011 1v2a1 1 0 001 1h2a1 1 0 001-1v-6.586l.293.293a1 1 0 001.414-1.414l-7-7z"
                      />
                    </svg>
                    <span
                      class="sr-only"
                    >
                      Home
                    </span>
                  </a>
                </div>
              </li>
              <li>
                <div
                  class="flex items-center"
                >
                  <svg
                    aria-hidden="true"
                    class="flex-shrink-0 w-5 h-5 text-gray-300"
                    fill="currentColor"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
                      fill-rule="evenodd"
                    />
                  </svg>
                  <a
                    class="ml-4 text-sm font-medium text-gray-400 hover:text-gray-700"
                    href="#"
                  >
                    Supplier List
                  </a>
                </div>
              </li>
              <li>
                <div
                  class="flex items-center"
                >
                  <svg
                    aria-hidden="true"
                    class="flex-shrink-0 w-5 h-5 text-gray-300"
                    fill="currentColor"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
                      fill-rule="evenodd"
                    />
                  </svg>
                  <a
                    aria-current="page"
                    class="ml-4 text-sm font-medium text-gray-400 hover:text-gray-700"
                    href="#"
                  >
                    Supplier A
                  </a>
                </div>
              </li>
            </ol>
          </nav>
        </div>
      </header>
      <div
        class="flex items-stretch flex-1 min-h-full overflow-hidden"
      >
        <main
          class="flex-1 overflow-y-auto"
        >
          <div
            class="p-5 max-w-7xl"
          >
            <h1
              class="text-2xl font-semibold text-gray-900"
            >
              Dashboard
            </h1>
          </div>
          <section
            class="flex flex-col flex-1 min-w-0 p-3 mx-5 text-gray-600 bg-white rounded lg:order-last"
          >
            <p>
              children
            </p>
          </section>
        </main>
      </div>
    </div>
  </div>
</div>
`;
